### PR TITLE
Nd field basic tests2

### PR DESCRIPTION
**Pull request by fangohr**
Same as #258, but without the last two (accidental) merges.

### CODE REVIEW

**Comment by fangohr** [comment]
(i) The `test.docs` tests fail - the docstrings are outdated (due to change of output to arrays I think).

**Comment by fangohr** [comment]
(ii) The method in `@_as_array.register(Field)` method (https://github.com/ubermag/discretisedfield/pull/271/files#diff-c58c98f0ea2639669e6b8e97bd7e9eedc3cb071080df2e78678b2e70fe76cdd4R3614) seems to be specific to 3d fields?

**Comment by fangohr** [comment]
(iii) the documentation string at https://github.com/ubermag/discretisedfield/blob/0a2274ecfc1e4dcd6b766542ebef823884eb65f8/discretisedfield/field.py#L236 refers to nvdim=1 and nvdim=3 . Are other dimensions allowed? Probably needs updating.

**Comment by fangohr** [comment]
(iv) The documentation of `Mesh.index2point` (https://github.com/ubermag/discretisedfield/pull/271/files#diff-c61c4320b3a6f255741f90a4610a5b2aae26f9698465b9fabab7e35b43b3938eR656) is restricted to 3d only.

**Comment by fangohr** [comment]
(v) documentation for `Mesh.point2index` is restricted to 3d meshes (https://github.com/ubermag/discretisedfield/pull/271/files#diff-c61c4320b3a6f255741f90a4610a5b2aae26f9698465b9fabab7e35b43b3938eR717)

File: discretisedfield/tests/test_mesh.py [hunk 1]
```diff
@@ -636,10 +743,22 @@ def test_index2point_point2index_mutually_inverse():
...
 def test_region2slice():
+    p1 = 0
+    p2 = 4
+    cell = 1
+    mesh = df.Mesh(p1=p1, p2=p2, cell=cell)
+    assert isinstance(mesh, df.Mesh)
+    assert mesh.region2slices(df.Region(p1=p1, p2=p2)) == (slice(0, 4, None),)
+    assert mesh.region2slices(df.Region(p1=0, p2=1)) == (slice(0, 1, None),)
+    assert mesh.region2slices(df.Region(p1=2, p2=3)) == (slice(2, 3, None),)
+
```
Comment (fangohr) [inline discretisedfield/tests/test_mesh.py:758]:
```suggestion
    # check error is raised if region is greater than the mesh's region
```

File: discretisedfield/tests/test_mesh.py [hunk 2]
```diff
@@ -636,10 +743,22 @@ def test_index2point_point2index_mutually_inverse():
         assert np.allclose(mesh.index2point(mesh.point2index(p)), p, atol=0)
 
     for i in [(1, 0, 0), (0, 1, 0), (0, 0, 1), (1, 1, 1)]:
-        assert mesh.point2index(mesh.index2point(i)) == i
+        assert all(mesh.point2index(mesh.index2point(i)) == i)
 
 
 def test_region2slice():
+    p1 = 0
```
Comment (fangohr) [inline discretisedfield/tests/test_mesh.py:750]: Should we have a test for a 4d domain (i.e. ndim=4)? (or any dimension > 3)

File: discretisedfield/tests/test_mesh.py [hunk 3]
```diff
@@ -674,8 +803,25 @@ def test_points():
     assert np.allclose(mesh.points.y, [1.0, 3.0, 5.0], atol=0)
     assert np.allclose(mesh.points.z, [0.5, 1.5, 2.5, 3.5], atol=0)
 
+    p1 = (0, 0, 4, 4)
```
Comment (fangohr) [inline discretisedfield/tests/test_mesh.py:806]:
```suggestion
    # 4d example (ndim=4)
    p1 = (0, 0, 4, 4)
```

File: discretisedfield/tests/test_mesh.py [hunk 4]
```diff
@@ -663,8 +782,18 @@ def test_region2slice():
     with pytest.raises(ValueError):
         mesh.region2slices(df.Region(p1=(-1, 3, -1), p2=(3, 5, 0)))
 
+    with pytest.raises(ValueError):
+        mesh.region2slices(df.Region(p1=(-1, 3), p2=(3, 5)))
+
 
 def test_points():
+    p1 = 0
```
Comment (fangohr) [inline discretisedfield/tests/test_mesh.py:790]:
```suggestion
    # 1d example (ndim=1)
    p1 = 0
```

File: discretisedfield/tests/test_mesh.py [hunk 5]
```diff
@@ -663,8 +782,18 @@ def test_region2slice():
...
+    p2 = 10
+    cell = 2
+    mesh = df.Mesh(region=df.Region(p1=p1, p2=p2), cell=cell)
+
+    assert np.allclose(mesh.points.x, [1.0, 3.0, 5.0, 7.0, 9.0], atol=0)
+
```
Comment (fangohr) [inline discretisedfield/tests/test_mesh.py:796]:
```suggestion
    
    # 3d example (ndim=3)
```

File: discretisedfield/tests/test_mesh.py [hunk 6]
```diff
@@ -1006,86 +1241,111 @@ def test_k3d_mpl_subregions(tmp_path):
 
 
 def test_scale():
-    p1 = (-50e-9, -50e-9, 0)
-    p2 = (50e-9, 50e-9, 20e-9)
-    cell = (1e-9, 1e-9, 2e-9)
+    mesh = df.Mesh(p1=0, p2=10, cell=1, subregions={"sr": df.Region(p1=1, p2=5)})
```
Comment (fangohr) [inline discretisedfield/tests/test_mesh.py:1244]:
```suggestion
    # example 1: 1d mesh, 1 subregion
    mesh = df.Mesh(p1=0, p2=10, cell=1, subregions={"sr": df.Region(p1=1, p2=5)})
```

File: discretisedfield/tests/test_mesh.py [hunk 7]
```diff
@@ -1006,86 +1241,111 @@ def test_k3d_mpl_subregions(tmp_path):
...
+    assert np.allclose(res.region.pmin, -7.5, atol=0)
+    assert np.allclose(res.region.pmax, 17.5, atol=0)
+    assert np.allclose(res.region.edges, 25, atol=0)
+    assert np.all(res.n == n)
+    assert np.allclose(res.cell, 2.5, atol=0)
+    assert len(res.subregions) == 1
+    assert np.allclose(res.subregions["sr"].pmin, -5, atol=0)
+    assert np.allclose(res.subregions["sr"].pmax, 5, atol=0)
+
+    mesh = df.Mesh(p1=0, p2=10, cell=1, subregions={"sr": df.Region(p1=1, p2=5)})
```
Comment (fangohr) [inline discretisedfield/tests/test_mesh.py:1258]:
```suggestion
    # example 2: 1d mesh, 1 subregion, with given reference_point
    mesh = df.Mesh(p1=0, p2=10, cell=1, subregions={"sr": df.Region(p1=1, p2=5)})
```

File: discretisedfield/tests/test_mesh.py [hunk 8]
```diff
@@ -1006,86 +1241,111 @@ def test_k3d_mpl_subregions(tmp_path):
...
+    assert np.allclose(mesh.region.pmin, 6.3, atol=0)
+    assert np.allclose(mesh.region.pmax, 31.3, atol=0)
+    assert np.allclose(mesh.region.edges, 25, atol=0)
+    assert np.all(mesh.n == n)
+    assert np.allclose(mesh.cell, 2.5, atol=0)
+    assert len(mesh.subregions) == 1
+    assert np.allclose(mesh.subregions["sr"].pmin, 8.8, atol=0)
+    assert np.allclose(mesh.subregions["sr"].pmax, 18.8, atol=0)
+
+    p1 = (-50e-9, 0)
```
Comment (fangohr) [inline discretisedfield/tests/test_mesh.py:1272]:
```suggestion
    # example 3: 2d mesh, no subregions
    p1 = (-50e-9, 0)
```

**Review by fangohr — COMMENTED** [review]
Overall standard is very good. I like the great exploitation of pytest as well.

File: discretisedfield/mesh.py [hunk 9]
```diff
@@ -773,9 +807,9 @@ def region2slices(self, region):
         (slice(0, 10, None), slice(0, 5, None), slice(0, 1, None))
```
[the remark this answers is not in the document]
Comment (lang-m) [reply discretisedfield/mesh.py:807]: The `None` refers to the step (1 if not specified otherwise).

**Review by lang-m — APPROVED** [review]

**Pull request merged**